### PR TITLE
DS-126 Update patch on Dask-ml v1.3.0

### PR DESCRIPTION
**Pull request by jamesbsilva**
NOTE:
A) Ignore for this PR all the commits not made by myself as we probably dont have time to code review dask-ml v1.3.0
B) Please ignore the code I havent changed ie (the official dask code). I agree there are issues that you can code review on how things are but outside of C) please focus on the "changes and functionality of the patch" as the scope not what the official dask code could do better although I did try to improve some variable names and documentation when possible.
C) If you believe there is a bug in the official dask code please note it in this PR and will need another PR to deal with bug fix that should probably be passed on to the official dask-ml repo.

This PR makes the patch compatible with dask-ml in its current state. It adds the patch's functionality of applying the weighting scheme to the test fold. It does add 3 bits of extra functionaility.

support for eval_sample_weight to allow for a different weighting scheme on evaluation than training
support lightGBM eval_data_set by resolving the proper test fold partitions if the user provides the eval_data_set
new test case for sklearn.pipeline . This is used by ZEFR so should probably tested as well.
Files Changed to add functionality :

tests/model_selection/dask_searchcv/test_model_selection.py
dask_ml/model_selection/methods.py
dask_ml/model_selection/_search.py

### CODE REVIEW

File: dask_ml/model_selection/_search.py [hunk 1]
```diff
@@ -286,16 +304,50 @@ def normalize_params(params):
...
+
+    # sklearn subscripting adds __ scores to step variables
+    for candidate_param in fit_params:
+        if "eval_sample_weight" in candidate_param:
+            eval_weight_source = candidate_param
+            break
+        elif "sample_weight" in candidate_param:
+            eval_weight_source = candidate_param
+            break
+    return eval_weight_source
```
Comment (tian-yi-zefr) [inline dask_ml/model_selection/_search.py]:
```
candidate_set = set(candidate_param.keys())
return  "eval_sample_weight" in candidate_set or "sample_weight" in candidate_set or None
```
you may simplify this function like above

File: dask_ml/model_selection/methods.py [hunk 2]
```diff
@@ -549,7 +519,6 @@ def create_cv_results(
                 n_splits,
                 n_candidates,
                 splits=True,
```
Comment (tian-yi-zefr) [inline dask_ml/model_selection/methods.py:531]: any reason why

File: dask_ml/model_selection/methods.py [hunk 3]
```diff
@@ -549,7 +519,6 @@ def create_cv_results(
...
-                weights=weights
```
Comment (tian-yi-zefr) [inline dask_ml/model_selection/methods.py:552]: any reason why `weights` parameter is removed?

File: dask_ml/model_selection/_search.py [hunk 4]
```diff
@@ -286,16 +304,50 @@ def normalize_params(params):
...
+        cv: cv  dask task name,
+        fit_params: keys of fit params in (name, full_name) format
+        vals: values of fit params
+        n_splits: fold index
+        keys_filtered:
+    '''
+    if not fit_params:
+        return [(n, (None,None)) for n in range(n_splits)]
+    keys, vals = _generate_fit_params_key_vals(fit_params, keys_filtered=keys_filtered)
+    return [(n, ( (cv_extract_params, cv, keys, vals, n, True), (cv_extract_params, cv, keys, vals, n, False)) ) for n in range(n_splits)]
```
Comment (tian-yi-zefr) [inline dask_ml/model_selection/_search.py]: why do we need both `True` and `False`?

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 5]
```diff
@@ -975,21 +976,23 @@ def test_mock_with_fit_param_raises():
...
-    (brier_score_loss, False, True, [2500000, 500000, 200000, 100000]),
+@pytest.mark.parametrize("metric,greater_is_better,needs_proba,estimator,estimator_params,sample_wt", [
+    (accuracy_score, True, False, LogisticRegression(), {"random_state": [15432], "solver": ["lbfgs"]}, [1, 999999, 1, 999999]),
+    (accuracy_score, True, False, LogisticRegression(), {"random_state": [15432], "solver": ["lbfgs"]}, [100000, 200000, 100000, 200000]),
+    (accuracy_score, True, False, LogisticRegression(), {"random_state": [15432], "solver": ["lbfgs"]}, [100000, 100000, 100000, 100000]),
+    (accuracy_score, True, False, LogisticRegression(), {"random_state": [15432], "solver": ["lbfgs"]}, [200000, 100000, 200000, 100000]),
+    (accuracy_score, True, False, LogisticRegression(), {"random_state": [15432], "solver": ["lbfgs"]}, [999999, 1, 999999, 1]),
+    (accuracy_score, True, False, LogisticRegression(), {"random_state": [15432], "solver": ["lbfgs"]}, [2000000, 1000000, 1, 999999]),
+    (accuracy_score, True, False, Pipeline([("clf", LogisticRegression())]), {"clf__random_state": [15432], "clf__solver": ["lbfgs"]}, [2000000, 1000000, 1, 999999]),
```
Comment (IanQS) [inline tests/model_selection/dask_searchcv/test_model_selection.py]: Cool stuff! Would you mind adding s comment as to why you're doing a `Pipeline` ? I'm guessing it's to test that the Pipeline works with dask search cv but might be useful

File: dask_ml/model_selection/_search.py (hunk 4 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: This CV extra function get train folds  weights from if `True` and  test fold weights when `False`

File: dask_ml/model_selection/methods.py (hunk 3 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/methods.py:552]: In the way it works for the train folds the weights are passed around in the `fit_params` object which is then extracted with the CV object that has the cache. This CV object will get the train weights from `fit_params` and with this patch it also get the test weights too.

TLDR; fit_params moves around the training process so the weights are fetched from that object instead of passing around the weights explicitly

File: dask_ml/model_selection/_search.py (hunk 1 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: will add in next commit.

File: dask_ml/model_selection/methods.py (hunk 3 above)
Comment (tian-yi-zefr) [reply dask_ml/model_selection/methods.py:552]: good idea!

File: dask_ml/model_selection/_search.py (hunk 1 above)
Comment (tian-yi-zefr) [reply dask_ml/model_selection/_search.py]:
```
candidate_set = set(candidate_param.keys())
if "eval_sample_weight" in candidate_set:
    return "eval_sample_weight"
elif "sample_weight" in candidate_set:
    return "sample_weight"
else:
    return None
```
the original code will return bool or None, this one returns str or None

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 6]
```diff
@@ -1022,15 +1025,20 @@ def train(y, norm1_wt, needs_proba):
 
     # There's nothing special about GridSearchCV. Could be RandomizedSearchCV.
     gscv = dcv.GridSearchCV(
-        LogisticRegression(),
-        {"random_state": [15432], "solver": ["lbfgs"]},  # parameters
+        estimator,
```
Comment (tian-yi-zefr) [inline tests/model_selection/dask_searchcv/test_model_selection.py:1042]: like the test case change!

File: dask_ml/model_selection/_search.py [hunk 7]
```diff
@@ -399,10 +440,18 @@ def do_fit_and_score(
...
+
         for n in range(n_splits):
-            train_sample_weight, test_sample_weight = _get_fold_sample_weights(
-                sample_weight, cv, n
-            )
+            if eval_weight_source is not None:
+                # format to keys with full information compatible with cv_extract functions
+                keys, vals = _generate_fit_params_key_vals(fit_params, keys_filtered=[eval_weight_source])
+                # dask evaluation requires wrapping into function to allow the function to be evaluated once cv object is resolved
+                def extract_param(cvs,k,v,n,fld):
```
Comment (tian-yi-zefr) [inline dask_ml/model_selection/_search.py]: move `extract_param` out of the for loop, it does not need to be defined multiple times.

**Comment by jamesbsilva** [comment]
FYI: This might change the current notebook behavior because of how test cases run on prod

http://dask.zefr.com:8888/lab/workspaces/james-chr/tree/mnt/efs/notebooks/james.silva/Hotfix/run_test_cases_on_prod_build.ipynb

File: dask_ml/model_selection/_search.py [hunk 8]
```diff
@@ -155,7 +153,15 @@ def decompress_params(fields, params):
         cache_cv=cache_cv,
     )
     cv_name = "cv-split-" + main_token
-    if iid:
+
+    ## DO WE NEED TO ADD SUPPORT FOR TPOP AND PATCH?
```
Comment (IanQS) [inline dask_ml/model_selection/_search.py:188]: Sorry, it looks like you're repeating code. Can you 

1) refactor to remove repetition

2) Add comments about what's happening? It's a little hard to review code when I'm not sure about the context. From my POV it looks like

File: dask_ml/model_selection/_search.py [hunk 9]
```diff
@@ -286,16 +304,50 @@ def normalize_params(params):
 
     return fields, tokens, params2
 
-
-def _get_fit_params(cv, fit_params, n_splits):
-    if not fit_params:
-        return [(n, None) for n in range(n_splits)]
+def _generate_fit_params_key_vals(fit_params, keys_filtered=None):
```
Comment (IanQS) [inline dask_ml/model_selection/_search.py:320]: if it's not too much of a pain can you add typing? I'm assuming `fit_params` is a dictionary but what is `keys_filtered` ? And what does it do?

File: dask_ml/model_selection/_search.py [hunk 10]
```diff
@@ -286,16 +304,50 @@ def normalize_params(params):
...
+        if "eval_sample_weight" in candidate_param:
+            eval_weight_source = candidate_param
+            break
+        elif "sample_weight" in candidate_param:
+            eval_weight_source = candidate_param
+            break
+    return eval_weight_source
+
+
+def _get_n_folds_fit_params(cv, fit_params, n_splits, keys_filtered=None):
```
Comment (IanQS) [inline dask_ml/model_selection/_search.py:387]: Can we get typing on this? Also, for the return type can we get a NamedTuple? It's a little hard to parse the return here. Would also make the below easier to read 

```
for n, fld_fit_params in n_and_fold_fit_params:
    ...
    ...
    fld_fit_params[0],
    fld_fit_params[1],
```

and it's used in other places too so I think returning a NamedTuple would be really beneficial

**Review by IanQS — COMMENTED** [review]
Generally looks good BUT I think that there are some low-hanging fruit that would be really useful for readability and maintainability moving forward. Happy to approve after

File: dask_ml/model_selection/_search.py (hunk 1 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: after looking at this it won't work since it won't give the right value for a pipeline

In a pipeline like `Pipeline([("clf", LogisticRegression())])` the `sample_weight` for the logistic regression will be `clf__sample_weight`  so it has to return  `clf__sample_weight` when it is that param

File: tests/model_selection/dask_searchcv/test_model_selection.py (hunk 5 above)
Comment (jamesbsilva) [reply tests/model_selection/dask_searchcv/test_model_selection.py]: Our use case in `ds-model-content-relevancy` uses `Pipeline` in practice so it is actually the most relevant for our use case. I am assuming that if I just put the `Logistic Regression` in a `pipeline` all the weighting is supposed to work exactly the same as if it was outside the the `pipeline`

File: dask_ml/model_selection/_search.py (hunk 9 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py:320]: I could add typing to the docstring for sure but I am avoiding making these 3 files the only files with typing in the whole repo.

File: dask_ml/model_selection/_search.py (hunk 8 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py:188]: 1) Most of the code reuse from `build_graph` and `build_cv_graph` is from the v1.3.0 base. From the one comment they added the plan is to deprecate `build_graph` once `TPOP` updates to be compatible with `dask-ml>=0.13.0`

2) I will add more typing in doc string but as an approximate rule generally if it is in `_search.py` the objects are actually Dask tasks and if it is in `methods.py` it has been computed and is a more concrete object post-compute.

File: dask_ml/model_selection/_search.py (hunk 10 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py:387]: see above comment. will add best I can in dostrings

File: dask_ml/model_selection/_search.py [hunk 11]
```diff
@@ -155,7 +153,15 @@ def decompress_params(fields, params):
...
     )
     cv_name = "cv-split-" + main_token
-    if iid:
+
+    ## DO WE NEED TO ADD SUPPORT FOR TPOP AND PATCH?
+    eval_weight_source = _get_weights_source(fit_params)
+    if eval_weight_source is not None :
+        weights = "cv-n-weights-" + main_token
+        # get test fold weights for all folds
+        fold_weights = [fld[1][1] for fld in _get_n_folds_fit_params(cv_name, fit_params, n_splits, keys_filtered=[eval_weight_source])]
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py]: Break this comprehension out into many lines.

File: dask_ml/model_selection/_search.py [hunk 12]
```diff
@@ -224,6 +230,18 @@ def build_cv_graph(
     cv_name = "cv-split-" + main_token
     dsk[cv_name] = (cv_split, cv, X_name, y_name, groups_name, is_pairwise, cache_cv)
 
+    eval_weight_source = _get_weights_source(fit_params)
+    if eval_weight_source is not None :
+        weights = "cv-n-weights-" + main_token
+        # get test fold weights for all folds
+        fold_weights = [fld[1][1] for fld in _get_n_folds_fit_params(cv_name, fit_params, n_splits, keys_filtered=[eval_weight_source])]
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py]: Break this comprehension out into many lines.

File: dask_ml/model_selection/_search.py (hunk 9 above)
Comment (ryan-deak-zefr) [reply dask_ml/model_selection/_search.py:320]: Yeah.  I would just add comments rather than typing to "keep it consistent with Dask style".  I know what this sounds like.  Obviously, comments are not in "Dask style", but comments are good so maybe just keep it "Dask(ish) style".

File: dask_ml/model_selection/_search.py [hunk 13]
```diff
@@ -286,16 +304,50 @@ def normalize_params(params):
...
+        if keys_filtered is not None and name not in keys_filtered:
+            continue
         vals.append(val)
         keys.append((name, full_name))
-    return [(n, (cv_extract_params, cv, keys, vals, n)) for n in range(n_splits)]
+    return keys, vals
+
+def _get_weights_source(fit_params):
+    '''
+        _get_weights_source returns the weights source. sklearn pipelines subscript the parameter names such that a "samples_weight" parameter must be resolved
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py]: Don't be afraid of `"\n"`.  Make these comments <= 78 characters long if possible.

File: dask_ml/model_selection/_search.py [hunk 14]
```diff
@@ -286,16 +304,50 @@ def normalize_params(params):
...
+            break
+        elif "sample_weight" in candidate_param:
+            eval_weight_source = candidate_param
+            break
+    return eval_weight_source
+
+
+def _get_n_folds_fit_params(cv, fit_params, n_splits, keys_filtered=None):
+    '''
+        _get_n_folds_fit_params gets index given by fold number and the fit_params for that folds train folds and test fold
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py:389]: add newlines.

File: dask_ml/model_selection/_search.py (hunk 4 above)
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py]: add newlines.

File: dask_ml/model_selection/_search.py [hunk 15]
```diff
@@ -399,10 +440,18 @@ def do_fit_and_score(
...
+        w_train, w_test = None, None
+
         for n in range(n_splits):
-            train_sample_weight, test_sample_weight = _get_fold_sample_weights(
-                sample_weight, cv, n
-            )
+            if eval_weight_source is not None:
+                # format to keys with full information compatible with cv_extract functions
+                keys, vals = _generate_fit_params_key_vals(fit_params, keys_filtered=[eval_weight_source])
+                # dask evaluation requires wrapping into function to allow the function to be evaluated once cv object is resolved
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py]: newlines

File: dask_ml/model_selection/_search.py [hunk 16]
```diff
@@ -399,10 +440,18 @@ def do_fit_and_score(
...
+                def extract_param(cvs,k,v,n,fld):
+                    return cvs.extract_param(k,v,n,fld)
+                w_train, w_test = (extract_param, cv, keys[0], vals[0], n,True), (extract_param, cv, keys[0], vals[0], n,False)
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py]: newlines

File: dask_ml/model_selection/_search.py [hunk 17]
```diff
@@ -1275,12 +1323,25 @@ def fit(self, X, y=None, groups=None, **fit_params):
         else:
             out = scheduler(dsk, keys, num_workers=n_jobs)
 
-        if self.iid:
+        distribution_warning = ' No explicit "eval_sample_weight" using sample_weights (if available or) equal / no weights. ' \
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py]: I would put this in parentheses so that you don't need that slashes and make these lines 78 chars or less.

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 18]
```diff
@@ -1022,15 +1025,20 @@ def train(y, norm1_wt, needs_proba):
...
 
-    gscv.fit(X, y, sample_weight=sample_weight)
+    if isinstance(estimator, Pipeline):
+        gscv.fit(X, y, clf__sample_weight=sample_weight)
+    else:
+        gscv.fit(X, y, sample_weight=sample_weight)
+
     best_score = gscv.best_score_
+
     np.testing.assert_almost_equal(exp_cv_score, best_score, decimal=5)
```
Comment (ryan-deak-zefr) [inline tests/model_selection/dask_searchcv/test_model_selection.py]: I think that `decimal` should be as high as possible.  I think that you should run this locally and keep incrementing it by one until it fails.  Then decrement and that should be the value.  I would extract this value to a constant at the top of the test and put a comment about how the logistic regression should be accurate to that value.

**Comment by jamesbsilva** [comment]
FYI:
I will add more typing in doc string but as an approximate rule generally if it is in `_search.py` the objects are actually Dask tasks (tuples with information how to calculate the task) and if it is in `methods.py` it has been computed and is a more concrete object post-compute like a bumpy array ect.

TLDR; **`_search.py` is pre-compute() and `methods.py` is mid-compute()**

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 19]
```diff
@@ -975,21 +976,23 @@ def test_mock_with_fit_param_raises():
...
-    (accuracy_score, True, False, [1, 999999, 1, 999999]),
-    (accuracy_score, True, False, [100000, 200000, 100000, 200000]),
-    (accuracy_score, True, False, [100000, 100000, 100000, 100000]),
-    (accuracy_score, True, False, [200000, 100000, 200000, 100000]),
-    (accuracy_score, True, False, [999999, 1, 999999, 1]),
-    (accuracy_score, True, False, [2000000, 1000000, 1, 999999]),
-    (log_loss, False, True, [2500000, 500000, 200000, 100000]),
-    (brier_score_loss, False, True, [2500000, 500000, 200000, 100000]),
+@pytest.mark.parametrize("metric,greater_is_better,needs_proba,estimator,estimator_params,sample_wt", [
+    (accuracy_score, True, False, LogisticRegression(), {"random_state": [15432], "solver": ["lbfgs"]}, [1, 999999, 1, 999999]),
```
Comment (ryan-deak-zefr) [inline tests/model_selection/dask_searchcv/test_model_selection.py]: I would create a `IMP_WT_LOG_REG_PARAMS` constant just above the test like the following:

```python
IMP_WT_LOG_REG_PARAMS = {
    "solver":         "lbfgs",
    "penalty":        "l2",
    "tol":            1e-6,
    "max_iter":       10000,
    "fit_intercept":  False,
    "random_state":   42
}
```

Then create the logistic regression models in the test cases as:

```python
LogisticRegression(**IMP_WT_LOG_REG_PARAMS)
```

This will provide many benefits like locking the random seed if any randomization is used, guarantee a fitting algorithm and attempt to make the model overfit as much as possible.  This will help ensure that the tests can increase the numeric precision in the calls to `np.testing.assert_almost_equal` in the test.

**Review by ryan-deak-zefr — CHANGES_REQUESTED** [review]

File: dask_ml/model_selection/methods.py [hunk 20]
```diff
@@ -162,13 +169,23 @@ def cv_n_samples(cvs):
...
+
 
+def cv_extract_params(cvs, keys, vals, n, is_train_folds):
+    '''
+        cv_extract_params the fit parameters of the fold sets (train folds or test fold)
 
-def cv_extract_params(cvs, keys, vals, n):
-    return {k: cvs.extract_param(tok, v, n) for (k, tok), v in zip(keys, vals)}
+        cvs: (CVCache): CV cache for CV information of folds
+        keys: ((str,str)) fit params (name,full_name) key tuple
```
Comment (jamesbsilva) [inline dask_ml/model_selection/methods.py:181]: I think having a key that is just the `full_name` or `name` of the fit parameter key is simpler but `v1.3.0` of dask-ml uses  this tuple as the key and I am avoiding changing this key format as I don't want to find out want to find out what the downstream impact is on this change for the sake of a simpler key scheme. This is also why `_generate_fit_params_key_vals` was necessary

File: dask_ml/model_selection/_search.py (hunk 13 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: will add in next commit

File: dask_ml/model_selection/_search.py (hunk 14 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py:389]: will add in next commit

File: tests/model_selection/dask_searchcv/test_model_selection.py (hunk 18 above)
Comment (jamesbsilva) [reply tests/model_selection/dask_searchcv/test_model_selection.py]: It looks like the same incrementing test must have happened in the original patch because it broke at 6 😄

File: dask_ml/model_selection/_search.py [hunk 21]
```diff
@@ -286,16 +304,65 @@ def normalize_params(params):
...
+
+        returns:
+            list( (int, tuple(Any) )
+            list of tuples of fold number and set of dask task that extracts that folds fit parameters for train folds and test folds
+            These tasks will be evaluated by scheduler later
+    '''
+    if not fit_params:
+        return [(n, (None,None)) for n in range(n_splits)]
+    keys, vals = _generate_fit_params_key_vals(fit_params, keys_filtered=keys_filtered)
+    return [(n, ( (cv_extract_params, cv, keys, vals, n, True), (cv_extract_params, cv, keys, vals, n, False)) ) for n in range(n_splits)]
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py]:
```suggestion
return [
    (
        n,
        (
            (cv_extract_params, cv, keys, vals, n, True),
            (cv_extract_params, cv, keys, vals, n, False)
        )
    ) for n in range(n_splits)
]
```

There's a couple of really dense, single line list comprehensions that need to be split up like this. 
Especially with the dense tuples improving readability here is important.

File: dask_ml/model_selection/_search.py (hunk 11 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: Broke out into more lines and added comment too

File: dask_ml/model_selection/_search.py (hunk 17 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: done

File: tests/model_selection/dask_searchcv/test_model_selection.py (hunk 19 above)
Comment (jamesbsilva) [reply tests/model_selection/dask_searchcv/test_model_selection.py]: added

File: dask_ml/model_selection/_search.py (hunk 21 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: Reformatted tuple and added comment on the stuff in tuple

File: dask_ml/model_selection/_search.py (hunk 7 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: `cv_extract_params` is slightly different than `extract_param`. `cv_extract_params` is the original Dask function that gets the fit_params information in `{fit_param_key: key_values}` format  for all the `fit_params` . `extract_param` is a one time pull of a single `fit_param` since it needs to be passed on to `score` for a pipeline before it has actually been computed.

File: dask_ml/model_selection/_search.py [hunk 22]
```diff
@@ -286,16 +308,77 @@ def normalize_params(params):
...
+        n_splits: int: fold index
+        keys_filtered: list[str] keys to filter from fit_params from output
+
+        returns:
+            list( (int, tuple(Any) )
+            list of tuples of fold number and set of dask task that extracts that folds fit parameters for train folds and test folds
+            These tasks will be evaluated by scheduler later
+    '''
+    if not fit_params:
+        return [(n, (None,None)) for n in range(n_splits)]
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py:402]: New line after indented block.

File: dask_ml/model_selection/_search.py [hunk 23]
```diff
@@ -423,8 +507,8 @@ def do_fit_and_score(
                     ytrain,
                     scorer,
                     error_score,
-                    test_sample_weight,
-                    train_sample_weight,
+                    w_train,
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py:543]: Changed the ordering of train and test is this intended?

File: dask_ml/model_selection/_search.py [hunk 24]
```diff
@@ -286,16 +308,77 @@ def normalize_params(params):
...
+        returns: str of source of test folds weights in fit_params dictionary
+    '''
+    eval_weight_source = None
+    if fit_params is None:
+        return eval_weight_source
+
+    # sklearn subscripting adds __ scores to step variables
+    for candidate_param in fit_params:
+        if "eval_sample_weight" in candidate_param:
+            eval_weight_source = candidate_param
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py]: Why not just return `candidate_param`?

File: dask_ml/model_selection/_search.py [hunk 25]
```diff
@@ -286,16 +308,77 @@ def normalize_params(params):
...
+            break
+        elif "sample_weight" in candidate_param:
+            eval_weight_source = candidate_param
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py]: Same as above

File: dask_ml/model_selection/_search.py [hunk 26]
```diff
@@ -224,6 +232,20 @@ def build_cv_graph(
     cv_name = "cv-split-" + main_token
     dsk[cv_name] = (cv_split, cv, X_name, y_name, groups_name, is_pairwise, cache_cv)
 
+    eval_weight_source = _get_weights_source(fit_params)
+    if eval_weight_source is not None :
+        weights = "cv-n-weights-" + main_token
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py]: The code in this block is replicated exactly above at lines 160 to 165. Would it make sense to DRY this out and put it in a function?

File: dask_ml/model_selection/_search.py [hunk 27]
```diff
@@ -155,7 +153,17 @@ def decompress_params(fields, params):
...
-    if iid:
+
+    ## DO WE NEED TO ADD SUPPORT FOR TPOP AND PATCH?
+    eval_weight_source = _get_weights_source(fit_params)
+    if eval_weight_source is not None :
+        weights = "cv-n-weights-" + main_token
+        # get all folds fit params info
+        folds_fp = _get_n_folds_fit_params( cv_name, fit_params, n_splits, keys_filtered=[eval_weight_source])
+        # 1 index is the tuple of fit_params of train folds and test fold
+        test_fold_weights = [ fld[1][1] for fld in folds_fp ]
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py]: The test fold is always at index 1?

File: dask_ml/model_selection/_search.py (hunk 26 above)
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py]: Initialization of weights can be pulled outside of conditionals and then just replace with None in the else case.

File: dask_ml/model_selection/_search.py [hunk 28]
```diff
@@ -1275,12 +1359,29 @@ def fit(self, X, y=None, groups=None, **fit_params):
...
+
+        if eval_weight_source is not None or self.iid:
             weights = out[0]
             scores = out[1:]
+            # reduce weights in folds to support cross fold averaging
+            if eval_weight_source is not None:
+                weights = np.array([np.sum(x[eval_weight_source]) for x in weights])
+            # output distribution warning as suggested if eval_sample_weight is not explicitly provided
+            if not "eval_sample_weight" in eval_weight_source:
+                logger.warning(distribution_warning)
```
Comment (zachary-mcpher) [inline dask_ml/model_selection/_search.py]: Could you break up the log message to be specific to the case here and the case at 1384?

**Review by zachary-mcpher — CHANGES_REQUESTED** [review]
A few comments

File: dask_ml/model_selection/_search.py (hunk 28 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: split

File: dask_ml/model_selection/_search.py (hunk 27 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: yup

File: dask_ml/model_selection/_search.py (hunk 26 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py]: put into new function

File: dask_ml/model_selection/_search.py (hunk 23 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py:543]: no preference on my end but the rest assumes this order for now. I could flip them,

**Comment by jamesbsilva** [comment]
Screen Shot 2020-04-23 at 1 56 38 PM


Results of tests

File: dask_ml/model_selection/methods.py (hunk 2 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/methods.py:531]: I think this is just a bool for a private store function that just reminds the store function to create different keys for different splits

File: dask_ml/model_selection/_search.py (hunk 23 above)
Comment (zachary-mcpher) [reply dask_ml/model_selection/_search.py:543]: No preference either I was just make sure this was intended.

**Review by zachary-mcpher — APPROVED** [review]
LGTM

File: dask_ml/model_selection/_search.py [hunk 29]
```diff
@@ -286,16 +317,98 @@ def normalize_params(params):
...
+    '''
+        _check_fit_params_key_used checks a fit_parameter is actually available
+        and not None.
+
+        fp: Dict[str,Any] fit_params
+        key: str fit parameter name
+
+        return bool status of fit parameter
+    '''
+    if key in fp:
```
Comment (tian-yi-zefr) [inline dask_ml/model_selection/_search.py:350]: logic is a little bit complicated here, can you simply this function? what cases should return True and rest of cases should return False.

File: dask_ml/model_selection/_search.py (hunk 7 above)
Comment (tian-yi-zefr) [reply dask_ml/model_selection/_search.py]: I mean you can define `extract_param` before for loop.

File: dask_ml/model_selection/_search.py (hunk 29 above)
Comment (jamesbsilva) [reply dask_ml/model_selection/_search.py:350]: added more comments in latest commit

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 30]
```diff
@@ -975,21 +976,35 @@ def test_mock_with_fit_param_raises():
...
+    (accuracy_score, True, False, LogisticRegression(), IMP_WT_LOG_REG_PARAMS, [100000, 200000, 100000, 200000]),
+    (accuracy_score, True, False, LogisticRegression(), IMP_WT_LOG_REG_PARAMS, [100000, 100000, 100000, 100000]),
+    (accuracy_score, True, False, LogisticRegression(), IMP_WT_LOG_REG_PARAMS, [200000, 100000, 200000, 100000]),
+    (accuracy_score, True, False, LogisticRegression(), IMP_WT_LOG_REG_PARAMS, [999999, 1, 999999, 1]),
+    (accuracy_score, True, False, LogisticRegression(), IMP_WT_LOG_REG_PARAMS, [2000000, 1000000, 1, 999999]),
+    (accuracy_score, True, False, LOG_REG_PIPELINE, IMP_WT_PIPE_LOG_REG_PARAMS, [2000000, 1000000, 1, 999999]),
+    (log_loss, False, True, LogisticRegression(), IMP_WT_LOG_REG_PARAMS, [2500000, 500000, 200000, 100000]),
+    (brier_score_loss, False, True, LogisticRegression(), IMP_WT_LOG_REG_PARAMS, [2500000, 500000, 200000, 100000]),
 ])
 def test_sample_weight_cross_validation(
```
Comment (zexuan-zhou) [inline tests/model_selection/dask_searchcv/test_model_selection.py:1003]: Looks like this is zefr test case? If so can we leave a comment here so that it can be easily found later?

**Review by tian-yi-zefr — APPROVED** [review]
Generally, it looks good to me. One thing I would do is testing the dependency upgrade on ds-model-content-relevancy and make sure the upgrade works fine.

File: tests/model_selection/dask_searchcv/test_model_selection.py (hunk 30 above)
Comment (jamesbsilva) [reply tests/model_selection/dask_searchcv/test_model_selection.py:1003]: will do

**Review by zexuan-zhou — APPROVED** [review]
lgtm

File: dask_ml/model_selection/_search.py [hunk 31]
```diff
@@ -399,10 +503,23 @@ def do_fit_and_score(
...
         for n in range(n_splits):
-            train_sample_weight, test_sample_weight = _get_fold_sample_weights(
-                sample_weight, cv, n
-            )
+            if eval_weight_source is not None:
+                # format keys with full information compatible with cv_extract functions
+                keys, vals = _generate_fit_params_key_vals(fit_params, keys_filtered=[eval_weight_source])
+                # create the proper dask tasks to generate the train objects when computing.
+                # Dask tasks are tuples of function followed by arguments
+                w_train = (extract_param, cv, keys[0], vals[0], n, True)
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py:521]: Maybe a comment about what the `0` index is.

File: dask_ml/model_selection/_search.py [hunk 32]
```diff
@@ -491,7 +608,7 @@ def do_fit(
                         error_score,
                         fields,
                         p,
-                        fit_params,
+                        fld_fit_params[0],
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py:611]: Maybe a comment about what the `0` index is.

File: dask_ml/model_selection/_search.py [hunk 33]
```diff
@@ -578,7 +695,7 @@ def do_fit_transform(
...
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py:698]: Maybe a comment about what the `0` index is.

**Review by ryan-deak-zefr — APPROVED** [review]
The only thing I would like to see is comments about the meaning of indices when indexing into the dask keys and values and params arrays.
